File: Resources/gen_icon.swift
// Genera l'iconset per l'app Libreria, usando solo AppKit.
// Uso: swift gen_icon.swift <cartella_output>

import AppKit
import Foundation

let outputDir = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "."
let outputURL = URL(fileURLWithPath: outputDir)
try? FileManager.default.createDirectory(at: outputURL, withIntermediateDirectories: true)

func crea(size: Int, to file: URL) {
    let s = CGFloat(size)
    let image = NSImage(size: NSSize(width: s, height: s))
    image.lockFocus()

    guard let ctx = NSGraphicsContext.current?.cgContext else { image.unlockFocus(); return }
    ctx.setShouldAntialias(true)
    ctx.interpolationQuality = .high

    // -- sfondo squircle --
    let rect = NSRect(x: 0, y: 0, width: s, height: s)
    let radius = s * 0.225
    let bg = NSBezierPath(roundedRect: rect, xRadius: radius, yRadius: radius)
    NSColor(srgbRed: 0.545, green: 0.227, blue: 0.180, alpha: 1.0).setFill()
    bg.fill()

    // bordo interno leggermente più chiaro (effetto profondità)
    let inset = s * 0.015
    let innerRect = rect.insetBy(dx: inset, dy: inset)
    let inner = NSBezierPath(roundedRect: innerRect, xRadius: radius * 0.95, yRadius: radius * 0.95)
    NSColor(srgbRed: 0.62, green: 0.27, blue: 0.21, alpha: 1.0).setFill()
    inner.fill()

    // ombra molto leggera sotto i libri
    ctx.saveGState()
    ctx.setShadow(offset: CGSize(width: 0, height: -s * 0.01), blur: s * 0.015,
                  color: NSColor.black.withAlphaComponent(0.25).cgColor)

    // -- libri in pila --
    let centroX = s * 0.50
    let centroY = s * 0.42
    let larghezzaLibro = s * 0.13
    let altezzaLibro = s * 0.36
    let gap = s * 0.018

    struct Libro { var off: CGFloat; var r: CGFloat; var g: CGFloat; var b: CGFloat; var extra: CGFloat }
    let libri: [Libro] = [
        Libro(off: -2.2 * (larghezzaLibro + gap), r: 0.96, g: 0.92, b: 0.85, extra: 0.0),
        Libro(off: -1.1 * (larghezzaLibro + gap), r: 0.79, g: 0.27, b: 0.22, extra: 0.04),
        Libro(off:  0.0,                          r: 0.96, g: 0.92, b: 0.85, extra: 0.10),
        Libro(off:  1.1 * (larghezzaLibro + gap), r: 0.20, g: 0.40, b: 0.55, extra: 0.02),
        Libro(off:  2.2 * (larghezzaLibro + gap), r: 0.85, g: 0.70, b: 0.40, extra: 0.0),
    ]

    let baseY = centroY - altezzaLibro / 2
    let rLibro = s * 0.012

    for lib in libri {
        let x = centroX - larghezzaLibro / 2 + lib.off
        let h = altezzaLibro * (1.0 + lib.extra)
        let r = NSRect(x: x, y: baseY, width: larghezzaLibro, height: h)
        let bp = NSBezierPath(roundedRect: r, xRadius: rLibro, yRadius: rLibro)
        NSColor(srgbRed: lib.r, green: lib.g, blue: lib.b, alpha: 1.0).setFill()
        bp.fill()

        // bande decorative
        let bandaAlta = NSRect(x: x, y: baseY + h * 0.80, width: larghezzaLibro, height: h * 0.04)
        NSColor.black.withAlphaComponent(0.16).setFill()
        NSBezierPath(rect: bandaAlta).fill()

        let bandaBassa = NSRect(x: x, y: baseY + h * 0.10, width: larghezzaLibro, height: h * 0.025)
        NSColor.black.withAlphaComponent(0.12).setFill()
        NSBezierPath(rect: bandaBassa).fill()
    }

    ctx.restoreGState()

    image.unlockFocus()

    // estrai PNG
    guard let tiff = image.tiffRepresentation,
          let rep = NSBitmapImageRep(data: tiff),
          let png = rep.representation(using: .png, properties: [:]) else { return }
    try? png.write(to: file)
}

let sizes: [(String, Int)] = [
    ("icon_16x16.png", 16),
    ("icon_16x16@2x.png", 32),
    ("icon_32x32.png", 32),
    ("icon_32x32@2x.png", 64),
    ("icon_128x128.png", 128),
    ("icon_128x128@2x.png", 256),
    ("icon_256x256.png", 256),
    ("icon_256x256@2x.png", 512),
    ("icon_512x512.png", 512),
    ("icon_512x512@2x.png", 1024),
]

print("Genero icone:")
for (name, sz) in sizes {
    let file = outputURL.appendingPathComponent(name)
    crea(size: sz, to: file)
    print("  → \(name) (\(sz)×\(sz))")
}
print("Fatto.")

File: Resources/gen_icon.py
#!/usr/bin/env python3
"""Genera un'icona PNG 1024x1024 in stile macOS per l'app Libreria.
Usa solo le framework Quartz/AppKit incluse in macOS (PyObjC).
"""
import sys
from pathlib import Path

try:
    from AppKit import (
        NSImage, NSBitmapImageRep, NSPNGFileType, NSColor, NSBezierPath,
        NSGraphicsContext, NSAffineTransform, NSFont, NSFontAttributeName,
        NSForegroundColorAttributeName, NSAttributedString, NSMakeRect,
        NSCompositingOperationSourceOver, NSShadow, NSMakeSize,
    )
    from Cocoa import NSSize
except ImportError:
    print("Errore: PyObjC non disponibile. Su macOS è incluso di default in /usr/bin/python3.")
    sys.exit(1)


def crea_icona(size: int, output: Path):
    image = NSImage.alloc().initWithSize_(NSSize(size, size))
    image.lockFocus()

    ctx = NSGraphicsContext.currentContext()
    ctx.setShouldAntialias_(True)
    ctx.setImageInterpolation_(3)  # high

    # ----- Sfondo arrotondato (squircle-ish) -----
    radius = size * 0.225
    rect = NSMakeRect(0, 0, size, size)
    path = NSBezierPath.bezierPathWithRoundedRect_xRadius_yRadius_(rect, radius, radius)

    # gradient simulato a 3 layer
    NSColor.colorWithSRGBRed_green_blue_alpha_(0.345, 0.118, 0.090, 1.0).setFill()
    path.fill()

    inset = size * 0.02
    inner = NSMakeRect(inset, inset, size - 2*inset, size - 2*inset)
    NSBezierPath.bezierPathWithRoundedRect_xRadius_yRadius_(inner, radius*0.95, radius*0.95).addClip()
    NSColor.colorWithSRGBRed_green_blue_alpha_(0.545, 0.227, 0.180, 1.0).setFill()
    path.fill()

    # ----- Libri in pila -----
    centro_x = size * 0.50
    centro_y = size * 0.42
    larghezza_libro = size * 0.13
    altezza_libro = size * 0.36
    gap = size * 0.018

    libri = [
        # (offset_x, color_r, color_g, color_b, altezza_extra)
        (-2.2 * (larghezza_libro + gap), 0.96, 0.92, 0.85, 0.0),   # crema
        (-1.1 * (larghezza_libro + gap), 0.79, 0.27, 0.22, 0.04),  # rosso
        ( 0.0,                            0.96, 0.92, 0.85, 0.08),  # crema (alto)
        ( 1.1 * (larghezza_libro + gap), 0.20, 0.40, 0.55, 0.02),   # blu
        ( 2.2 * (larghezza_libro + gap), 0.85, 0.70, 0.40, 0.0),    # ocra
    ]
    base_y = centro_y - altezza_libro / 2
    for offset_x, r, g, b, extra in libri:
        x = centro_x - larghezza_libro / 2 + offset_x
        h = altezza_libro * (1.0 + extra)
        libro_rect = NSMakeRect(x, base_y, larghezza_libro, h)
        radius_libro = size * 0.012
        bp = NSBezierPath.bezierPathWithRoundedRect_xRadius_yRadius_(libro_rect, radius_libro, radius_libro)
        NSColor.colorWithSRGBRed_green_blue_alpha_(r, g, b, 1.0).setFill()
        bp.fill()

        # banda decorativa nella parte alta
        banda_alta = NSMakeRect(x, base_y + h * 0.78, larghezza_libro, h * 0.04)
        bp_banda = NSBezierPath.bezierPathWithRect_(banda_alta)
        NSColor.colorWithSRGBRed_green_blue_alpha_(0, 0, 0, 0.15).setFill()
        bp_banda.fill()

        # banda in basso
        banda_bassa = NSMakeRect(x, base_y + h * 0.10, larghezza_libro, h * 0.025)
        bp_banda2 = NSBezierPath.bezierPathWithRect_(banda_bassa)
        NSColor.colorWithSRGBRed_green_blue_alpha_(0, 0, 0, 0.12).setFill()
        bp_banda2.fill()

    image.unlockFocus()

    # Estrai PNG
    tiff = image.TIFFRepresentation()
    rep = NSBitmapImageRep.imageRepWithData_(tiff)
    png_data = rep.representationUsingType_properties_(NSPNGFileType, {})
    png_data.writeToFile_atomically_(str(output), True)
    print(f"  → {output} ({size}×{size})")


if __name__ == "__main__":
    out_dir = Path(sys.argv[1]) if len(sys.argv) > 1 else Path(".")
    out_dir.mkdir(parents=True, exist_ok=True)

    print("Genero icone:")
    sizes = [
        ("icon_16x16.png", 16),
        ("icon_16x16@2x.png", 32),
        ("icon_32x32.png", 32),
        ("icon_32x32@2x.png", 64),
        ("icon_128x128.png", 128),
        ("icon_128x128@2x.png", 256),
        ("icon_256x256.png", 256),
        ("icon_256x256@2x.png", 512),
        ("icon_512x512.png", 512),
        ("icon_512x512@2x.png", 1024),
    ]
    for name, sz in sizes:
        crea_icona(sz, out_dir / name)
    print("Fatto.")
